_posts/2017-01-23-cpp-strings.md

```diff
@@ -0,0 +1,156 @@
+---
+title: Modeling strings with libpmemobj C++ bindings
+author: RobDickinson
+layout: post
+---
+
+C++ developers using libpmemobj have more than one option for modeling strings,
+depending on the size of the strings and whether they are fixed or varying
+in length. In this post we'll review the representations that work,
+known variations to avoid, and finally present a persistent string
+class that implements these best practices.
+
+#### Avoid wrapping fixed-size arrays
+
+You might expect (like I did at first!) that `p<char[size]>` is a proper way
+to simply model a fixed-size string, but actually this is not correct.
+The `p<>` template does not implement the subscript `operator[]` and so
+`p<char[size]>` won't even compile.
+
+#### Avoid wrapping std::string
+
+Never use `p<std::string>` or `persistent_ptr<std::string>`. This code will
+compile, but these implementations will not be power-fail safe (at best)
+and might be unstable (at worst). This advice applies to using `p<>`
+or `persistent_ptr<>` wrappers around any complex types that perform
+their own memory management.
+
+#### Avoid arrays of persistent pointers to single chars
+
+Don't model a variable-length string with representations like
+`persistent_ptr<char> X[]` or `persistent_ptr<p<char>[]>`. These are
+horribly inefficient and should always be avoided. Let's consider what
+happening behind the scenes here. Each char in the variable-length string
```

Comment:
Could you add that *In case of `persistent_ptr<char> X[]`, each char in the...*, because `persistent_ptr<p<char>[]>` will not have this overhead and I wouldn't want people to get confused.